patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.TML.cs

```
@@ -366,168 +282,225 @@ private void ReleaseWorkshopQuery()
-					return false;
+				foreach (var slug in queryParameters.searchModSlugs) {
+					// If Query Fails, we can't publish.
+					if (!TryRunQuery(SteamedWraps.GenerateModBrowserQuery(currentPage, queryParameters, internalName: slug))) {
+						ReleaseWorkshopQuery();
+						return false;
+					}
 
-				// If Query Succeeds, but doesn't find a match, assume safe to publish new item.
-				if (_queryReturnCount == 0)
-					return true;
+					if (_queryReturnCount == 0) {
+						items.Add(null);
+						ReleaseWorkshopQuery();
+						continue;
+					}
+
+					items.Add(GenerateModDownloadItemFromQuery(0));
+					ReleaseWorkshopQuery();
+				}
 
-				// Should only be one of the matching mod
-				itemDetails = SteamedWraps.FetchItemDetails(_primaryUGCHandle, 0);
 				return true;
 			}
 
-			internal bool TryRunQuery(SteamAPICall_t query)
+			/////// Run Queries ////////////////////
+
+			internal async Task<bool> TryRunQueryAsync(SteamAPICall_t query, CancellationToken token = default)
 			{
 				_primaryQueryResult = EResult.k_EResultNone;
 				_queryHook.Set(query);
 
-				var stopwatch = Stopwatch.StartNew();
+				Stopwatch stopwatch = Stopwatch.StartNew();
 				do {
-					if (stopwatch.Elapsed.TotalSeconds >= 10) // 10 seconds maximum allotted time before no connection is assumed
-						_primaryQueryResult = EResult.k_EResultTimeout;
-
-					SteamedWraps.ForceCallbacks();
+					await SteamedWraps.ForceCallbacks(token);
+				} while (
+					(_primaryQueryResult == EResult.k_EResultNone) &&
+					(stopwatch.Elapsed.TotalSeconds < 10)
+				);
+				if (_primaryQueryResult == EResult.k_EResultNone) {
+					_primaryQueryResult = EResult.k_EResultTimeout;
 				}
-				while (_primaryQueryResult == EResult.k_EResultNone);
 
+				// @TODO: HandleError calls show fancy page and UI stuff, is this ok since it is scheduled?
+				// Also shouldn't all the UI and handling happen in the browser and not in Workshop methods?
```

Solxan is entirely right here. You should throw an appropriate exception, async code is designed to propagate exceptions through to the awaiter, you should be able to flow them back to the call which asks for a query result on the (main or otherwise) thread. 

This method can be a `public static ReportWorkshopException(Exception)` message.